src/catalog/catalog_entry/dependency_set_catalog_entry.cpp

```diff
@@ -0,0 +1,223 @@
+	for (auto &dep : to_add) {
+		auto &entry = dep.entry;
+		AddDependency(transaction, entry, dep.dependency_type);
+	}
+}
+void DependencySetCatalogEntry::AddDependents(CatalogTransaction transaction, const dependency_set_t &to_add) {
+	for (auto &dep : to_add) {
+		auto &entry = dep.entry;
+		AddDependent(transaction, entry, dep.dependency_type);
+	}
+}
+
+// Add from a DependencyList
+void DependencySetCatalogEntry::AddDependencies(CatalogTransaction transaction, const DependencyList &to_add) {
+	for (auto &entry : to_add.set) {
+		AddDependency(transaction, entry);
+	}
+}
+void DependencySetCatalogEntry::AddDependents(CatalogTransaction transaction, const DependencyList &to_add) {
+	for (auto &entry : to_add.set) {
+		AddDependent(transaction, entry);
+	}
+}
+
+// Add from a single CatalogEntry
+void DependencySetCatalogEntry::AddDependency(CatalogTransaction transaction, CatalogEntry &to_add,
+                                              DependencyType type) {
+	static const DependencyList EMPTY_DEPENDENCIES;
+	auto dependency = make_uniq<DependencyCatalogEntry>(catalog, to_add, type);
+	auto dependency_name = dependency->name;
+	D_ASSERT(dependency_name != name);
+	if (catalog.IsTemporaryCatalog()) {
+		dependency->temporary = true;
+	}
+	dependencies.CreateEntry(transaction, dependency_name, std::move(dependency), EMPTY_DEPENDENCIES);
+}
+void DependencySetCatalogEntry::AddDependent(CatalogTransaction transaction, CatalogEntry &to_add,
+                                             DependencyType type) {
+	static const DependencyList EMPTY_DEPENDENCIES;
+	auto dependent = make_uniq<DependencyCatalogEntry>(catalog, to_add, type);
+	auto dependent_name = dependent->name;
+	D_ASSERT(dependent_name != name);
+	if (catalog.IsTemporaryCatalog()) {
+		dependent->temporary = true;
+	}
+	dependents.CreateEntry(transaction, dependent_name, std::move(dependent), EMPTY_DEPENDENCIES);
+}
+
+// Add from a Dependency
+void DependencySetCatalogEntry::AddDependency(CatalogTransaction transaction, Dependency to_add) {
+	AddDependency(transaction, to_add.entry, to_add.dependency_type);
+}
+void DependencySetCatalogEntry::AddDependent(CatalogTransaction transaction, Dependency to_add) {
+	AddDependent(transaction, to_add.entry, to_add.dependency_type);
+}
+
+static bool SkipDependencyRemoval(CatalogEntry &dependency) {
+	if (dependency.type != CatalogType::DEPENDENCY_ENTRY) {
+		return false;
+	}
+	auto &dep = dependency.Cast<DependencyCatalogEntry>();
+
+	// This link is not completed, so there is no dependency to remove
+	return dep.Type() == DependencyType::DEPENDENCY_OWNS;
+}
+
+// Remove dependency from a DependencyEntry
+void DependencySetCatalogEntry::RemoveDependency(CatalogTransaction transaction, CatalogEntry &dependency) {
+	D_ASSERT(dependency.type == CatalogType::DEPENDENCY_ENTRY || dependency.type == CatalogType::DEPENDENCY_SET);
+	if (SkipDependencyRemoval(dependency)) {
```

Comment:
This exists because for ALTER OWNERSHIP we create an uncompleted link, if we are aware of this we can verify that the link doesn't exist.
Problem is inside CleanupDependencies we don't pass in a DependencyCatalogEntry, so we're not aware that this is an OWNED BY relationship.

This is why we don't D_ASSERT that the DropEntry returned true.